File: Codes/src/javacode/No13904.java
package javacode;

import java.io.*;
import java.util.*;

public class No13904 {
    private static int n, result;
    private static boolean[] hasDate;
    private static PriorityQueue<Task> taskQueue = new PriorityQueue<>();

    public static void main(String[] args) throws IOException {
        input();
        for (int i = 0; i < n; i++) {
            Task task = taskQueue.poll();
            final int bestDate = findBestDate(task);
            if (bestDate == -1) {
                continue;
            }
            hasDate[bestDate] = true;
            result += task.score;
        }
        System.out.println(result);
    }
    private static int findBestDate(Task task) {
        for (int i = task.date; i > 0; i--) {
            if (!hasDate[i]) {
                return i;
            }
        }
        return -1;
    }
    private static void input() throws IOException {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(reader.readLine());
        int biggestDate = 0;
        for (int i = 0; i < n; i++) {
            StringTokenizer tokenizer = new StringTokenizer(reader.readLine());
            int date = Integer.parseInt(tokenizer.nextToken());
            int score = Integer.parseInt(tokenizer.nextToken());
            taskQueue.add(new Task(date, score));
            if (biggestDate < date) biggestDate = date;
        }
        hasDate = new boolean[biggestDate + 1];
    }

    static class Task implements Comparable<Task> {
        int date, score;
        public Task(int date, int score) {
            this.date = date;
            this.score = score;
        }

        public int compareTo(Task other) {
            return Integer.compare(other.score, this.score);
        }
    }
}

File: Codes/src/kotlincode/No13904.kt
import java.util.*

fun main() {
    val n = readln().toInt()
    val queue = PriorityQueue<Task>{one, other -> other.score.compareTo(one.score)}
    val hasDate: Array<Boolean> = Array(1001) { false }
    for (i in 0 until n) {
        val (dateString, scoreString) = readln().split(' ')
        queue.offer(Task(dateString.toInt(), scoreString.toInt()))
    }
    var sum = 0;
    while (!queue.isEmpty()) {
        sum += queue.poll().getScoreIfCanExecute(hasDate)
    }
    println(sum)
}

fun Task.findBestDate(hasDate: Array<Boolean>): Int {
    for (i in date downTo 1) {
        if (!hasDate[i]) {
            return i
        }
    }
    return -1
}

fun Task.getScoreIfCanExecute(hasDate: Array<Boolean>): Int = if (findBestDate(hasDate) != -1) {
    hasDate[findBestDate(hasDate)] = true
    score
} else 0

class Task(val date: Int, val score: Int)
